src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/Sources/GeneratedOutputSource.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Threading;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem.Sources;
+
+internal sealed class GeneratedOutputSource
+{
+    private readonly SemaphoreSlim _gate = new(initialCount: 1);
```

Comment:
In my opinion, this is fine for a couple of reasons:

1. Razor doesn't perform any blocking waits on the UI thread for a document's generated output.
2. The recommended replacement is MS.VS.Threading's `ReentrantSemaphore`, which requires a JTF to mitigate deadlocks. However, this layer shouldn't use JTF because it runs in many contexts where there isn't a main thread (e.g. Roslyn OOP and rzls).

Looking at the docs for `ReentrantSemaphore`, I think the only real benefit to us would be to create it with `ReentrancyMode.NotAllowed` so an exception is thrown if the same thread attempts to reenter the lock.